src/io.h

```diff
@@ -167,83 +167,363 @@ class FormatGRO {
+     * @param frame  target frame
+     * @throw std::runtime_error when the number of coordinates does not match
+     */
+    void exportFrame(TrajectoryFrame &frame) const;
+    /**
+     * @brief Exports data to scalar paramers and output iterator over atomic coordinates.
+     * @tparam begin_iterator
+     * @tparam end_iterator
+     * @param[out] step  frame step
+     * @param[out] time  frame timestamp
+     * @param[out] box  box dimensions (xyz) in nanometers
+     * @param[out] coordinates_begin  output iterator with coordinates in nanometers
+     * @param[out] coordinates_end  output iterator's end
+     * @throw std::runtime_error when the number of coordinates does not match
+     */
+    template <class begin_iterator, class end_iterator>
+    void exportFrame(int &step, float &time, Point &box, begin_iterator coordinates_begin,
+                     end_iterator coordinates_end) const {
+        exportTimestamp(step, time);
+        exportBox(box);
+        exportCoordinates(coordinates_begin, coordinates_end, 0.5 * box);
+    }
+
+  protected:
+    typedef float XTCFloat;
+    typedef Eigen::Matrix<XTCFloat, DIM, DIM, Eigen::RowMajor> XTCMatrix; //<! eigen equivalent of the box tensor
+    typedef Eigen::Matrix<XTCFloat, DIM, 1> XTCVector; //<! eigen equivalent of the single set of coordinates
+    /**
+     * @brief Imports and converts step and timestamp.
+     * @param[in] step  frame step
+     * @param[in] time  frame timestamp
+     */
+    void importTimestamp(const int step, const float time);
+    /**
+     * @brief Imports and converts simulation box dimensions.
+     * @param[in] box  simulation box dimensions in nanometers (xyz)
+     */
+    void importBox(const Point &box);
+    /**
+     * @brief Imports and converts atomic coordinates. Offset is added to all coordinates to account different
+     * coordinates' origin.
+     * @param[in] coordinates  atomic coordinates in nanometers
+     * @param[in] offset  offset in nanometers to add to all coordinates upon conversion
+     */
+    void importCoordinates(const PointVector &coordinates, const Point &offset);
+    /**
+     * @brief Imports and converts atomic coordinates. Offset is added to all coordinates to account different
+     * coordinates' origin.
+     * @tparam begin_iterator
+     * @tparam end_iterator
+     * @param[in] begin  input iterator with coordinates in nanometers
+     * @param[in] end  input iterator's end
+     * @param[in] offset  offset in nanometers to add to all coordinates upon conversion
+     */
+    template <class begin_iterator, class end_iterator>
+    void importCoordinates(begin_iterator begin, end_iterator end, const Point &offset) const {
+        // setNumberOfAtoms(std::distance(begin, end));
+        // comparison of i is probably faster than prior call of std::distance
+        int i = 0;
+        for (auto coordinates_it = begin; coordinates_it != end; ++coordinates_it) {
```

Comment:
use `std::transform` and `std::begin(xtc_coordinates.get())`. Throw can be moved outside loop.

Reply:
Does not look as straightforward with 2D C-arrays. Keeping the current code.